### ISSUE TITLE
[Bug]: Can't fold slash command output in chat buffer with Neovim Nightly

### ISSUE THREAD

**Issue by bassamsdata** [olimorris/codecompanion.nvim#190]
### Your `minimal.lua` config

```lua
local root = vim.fn.fnamemodify("./.repro", ":p")

-- set stdpaths to use .repro
for _, name in ipairs({ "config", "data", "state", "cache" }) do
  vim.env[("XDG_%s_HOME"):format(name:upper())] = root .. "/" .. name
end

-- set config path to enable the copilot adapter to work
vim.env["CODECOMPANION_TOKEN_PATH"] = vim.fn.expand("~/.config")

-- bootstrap lazy
local lazypath = root .. "/plugins/lazy.nvim"
if not vim.loop.fs_stat(lazypath) then
  vim.fn.system({
    "git",
    "clone",
    "--filter=blob:none",
    "--single-branch",
    "https://github.com/folke/lazy.nvim.git",
    lazypath,
  })
end
vim.opt.runtimepath:prepend(lazypath)

-- Your CodeCompanion setup
local plugins = {
  {
    "olimorris/codecompanion.nvim",
    dependencies = {
      { "nvim-treesitter/nvim-treesitter", build = ":TSUpdate"},
      "nvim-lua/plenary.nvim",
      "hrsh7th/nvim-cmp",
      {
        "stevearc/dressing.nvim",
        opts = {},
      },
      "nvim-telescope/telescope.nvim",
    },
    opts = {
      -- Refer to: https://github.com/olimorris/codecompanion.nvim/blob/main/lua/codecompanion/config.lua
      strategies = { -- Change the adapters as required
        chat = { adapter = "openai" },
        inline = { adapter = "openai" },
        agent = { adapter = "openai" },
      },
      opts = {
        log_level = "DEBUG",
      },
    },
  }
}

require("lazy").setup(plugins, {
  root = root .. "/plugins",
})

-- Setup Tree-sitter
local ok, treesitter = pcall(require, "nvim-treesitter.configs")
if ok then
  treesitter.setup({
    ensure_installed = { "lua", "markdown", "markdown_inline", "yaml" },
    highlight = { enable = true },
  })
end

-- Setup completion
local ok, cmp = pcall(require, "cmp")
if ok then
  cmp.setup({
    mapping = cmp.mapping.preset.insert({
      ["<C-b>"] = cmp.mapping.scroll_docs(-4),
      ["<C-f>"] = cmp.mapping.scroll_docs(4),
      ["<C-Space>"] = cmp.mapping.complete(),
      ["<C-e>"] = cmp.mapping.abort(),
      ["<CR>"] = cmp.mapping.confirm({ select = true }), -- Accept currently selected item. Set `select` to `false` to only confirm explicitly selected items.
    }),
  })
end
```

### Error messages

```lua
Error detected while processing InsertLeave Autocommands for "<buffer=3>":
Error executing lua callback: ...m-macos-x86_64/share/nvim/runtime/lua/vim/treesitter.lua:182: attempt to call method 'range' (a nil value)
stack traceback:
	...m-macos-x86_64/share/nvim/runtime/lua/vim/treesitter.lua:182: in function 'get_range'
	...m-macos-x86_64/share/nvim/runtime/lua/vim/treesitter.lua:217: in function 'get_node_text'
	...codecompanion.nvim/lua/codecompanion/strategies/chat.lua:1133: in function 'fold_code'
	...codecompanion.nvim/lua/codecompanion/strategies/chat.lua:1096: in function <...codecompanion.nvim/lua/codecompanion/strategies/chat.lua:1095>
```
<img width="1674" alt="Screenshot 2024-09-09 at 11 57 28 PM" src="https://github.com/user-attachments/assets/80539a69-7596-4bae-832d-37bf5a9b61bd">

### Log output

[DEBUG] 2024-09-09 23:57:15
Executing slash command: file
[DEBUG] 2024-09-09 23:57:15
Resolving slash command: helpers.slash_commands.file
[TRACE] 2024-09-10 00:01:41
Chat created with ID 7942063
[TRACE] 2024-09-10 00:01:41
Chat opened with ID 7942063
[TRACE] 2024-09-10 00:01:41
Setting the header for the chat buffer
[TRACE] 2024-09-10 00:01:41
Rendering headers in the chat buffer
[DEBUG] 2024-09-10 00:01:51
Executing slash command: file
[DEBUG] 2024-09-10 00:01:51
Resolving slash command: helpers.slash_commands.file
[DEBUG] 2024-09-10 00:02:21
Executing slash command: buffer
[DEBUG] 2024-09-10 00:02:21
Resolving slash command: helpers.slash_commands.buffer
[TRACE] 2024-09-10 00:04:15
Chat created with ID 7942063
[TRACE] 2024-09-10 00:04:15
Chat opened with ID 7942063
[TRACE] 2024-09-10 00:04:15
Setting the header for the chat buffer
[TRACE] 2024-09-10 00:04:15
Rendering headers in the chat buffer
[DEBUG] 2024-09-10 00:04:24
Executing slash command: file
[DEBUG] 2024-09-10 00:04:24
Resolving slash command: helpers.slash_commands.file

### Health check output

==============================================================================
codecompanion: require("codecompanion.health").check()

- Log file: /Users/bassam/.config/.repro/state/nvim/codecompanion.log

Plugins: ~
- OK plenary.nvim installed
- OK nvim-treesitter installed
- OK nvim-cmp installed
- OK telescope.nvim installed
- OK dressing.nvim installed

Tree-sitter parsers: ~
- OK markdown parser installed
- OK yaml parser installed

Libraries: ~
- OK curl installed

### Describe the bug

When using the /file or /buffer commands, an error appears after selecting the file or buffer. I believe this is related to the fold function, which uses Tree-sitter to detect the range.

I’m familiar with this bug and actually fixed the code for the `fold_code` function in the `chat.lua` file.

I have two solutions for this, and I can submit a PR if you're okay with it.. The first is simply a different approach to using Tree-sitter, but I’m unsure if it’s better or worse than the current function in terms of performance.
```lua
-- forgive the commented code, I just was trying things
function Chat:fold_code()
  if not self.has_folded_code then
    api.nvim_create_autocmd("InsertLeave", {
      group = self.aug,
      buffer = self.bufnr,
      desc = "Always fold code when a slash command is used",
      callback = function()
        self:fold_code()
      end,
    })
    self.has_folded_code = true
  end
  local parser = vim.treesitter.get_parser(self.bufnr, "markdown")
  local tree = parser:parse()[1]
  local root = tree:root()

  local function create_fold(start_row, end_row)
    vim.cmd(string.format("%d,%dfold", start_row + 1, end_row))
    -- vim.api.nvim_buf_set_extmark(self.bufnr, vim.api.nvim_create_namespace("codecompanion_folds"), start_row, 0, {
    --   virt_text = { { text, "Folded" } },
    --   virt_text_pos = "overlay",
    -- })
    return true
  end

  local function process_node(node)
    if node:type() == "paragraph" then
      local paragraph_text = vim.treesitter.get_node_text(node, self.bufnr)
      local file_name = paragraph_text:match("%[!File:%s*(.-)%]")
      local buffer = paragraph_text:match("%[!Buffer:%s*(.-)%]")
      if file_name or buffer then
        local next_node = node:next_sibling()
        if next_node and next_node:type() == "fenced_code_block" then
          local start_row, _, end_row, _ = next_node:range()
          -- local num_lines = end_row - start_row + 1
          -- local fold_text
          -- if file_name then
          --   fold_text = string.format("> File: %s (%d lines)", file_name, num_lines)
          -- else
          --   fold_text = string.format("> Buffer ID: %s (%d lines)", buffer, num_lines)
          -- end
          create_fold(start_row - 1, end_row)
        end
      else
      end
    end

    for child in node:iter_children() do
      process_node(child)
    end
  end

  process_node(root)
  vim.api.nvim_set_option_value("foldmethod", "manual", { win = self.winnr })
end
```
The other solution is to not rely on Tree-sitter at all. I’ve created a working function, but it still needs a bit of polishing.

```lua
function Chat:fold_code()
  if not self.has_folded_code then
    api.nvim_create_autocmd("InsertLeave", {
      group = self.aug,
      buffer = self.bufnr,
      desc = "Always fold code when a slash command is used",
      callback = function()
        self:fold_code()
      end,
    })
    self.has_folded_code = true
  end

  local lines = api.nvim_buf_get_lines(self.bufnr, 0, -1, false)
  local in_code_block = false
  local code_block_start = nil

  for i, line in ipairs(lines) do
    if line:match("^```") then
      if not in_code_block then
        in_code_block = true
        code_block_start = i
      else
        in_code_block = false
        if code_block_start then
          vim.cmd(string.format("%d,%dfold", code_block_start, i))
          code_block_start = nil
        end
      end
    elseif line:match("%[!File:") or line:match("%[!Buffer:") then
      if lines[i + 1] and lines[i + 1]:match("^```.*") then
        code_block_start = i
      end
    end
  end

  vim.api.nvim_buf_set_keymap(0, "n", "<tab>", "za", { noremap = true, silent = true })
  vim.api.nvim_set_option_value("foldmethod", "manual", { win = self.winnr })
end

```

### Reproduce the bug

1. CodeCompanionChat command
2. /file
3. select the file
4. issue appears

### Final checks

- [X] I have made sure this issue exists in the latest version of the plugin
- [X] I have tested with the `minimal.lua` file from above and have shared this
- [X] I have shared the contents of the log file

**Comment by olimorris** [comment]
I can't create this with the `minimal.lua` file. I suspect it must be a given file that you've used.

Could you try and create or share a file that causes this?

**Comment by bassamsdata** [comment]
> I can't create this with the minimal.lua file. I suspect it must be a given file that you've used.

My bad, that was on the Neovim nightly version. I forgot I’m using nightly.  
I still don't get folds in my config on Neovim 0.10.1, but that’s my issue to resolve.

I’m not sure if you want to keep this issue open for reference.

**Comment by olimorris** [comment]
Thanks for flagging. I periodically switch between Nightly and v0.10.1 so would have likely run into this. Although I only switched back about a week ago so they must have done something which affects us...

Let's leave this as open as it's definately an issue.

**Comment by olimorris** [comment]
Got it...Turns out the https://github.com/neovim/neovim/commit/6913c5e1d975a11262d08b3339d50b579e6b6bb8 PR broke the functionality. It's now resolved in 2.6.6. Can you test and let me know your verdict?

I thought it was odd because I'm subscribed to the breaking changes issue on nvim-treesitter. Lesson learned...watch [this](https://github.com/neovim/neovim/commits/98ba65b8be8cb2dde839502357e70916c4a3e37a/runtime/doc/news.txt) page for Nightly breaking changes.

**Comment by bassamsdata** [comment]
thank you. it worked like a charm.
